src/coreclr/vm/eetwain.cpp

```diff
@@ -4412,7 +4413,36 @@ void promoteVarArgs(PTR_BYTE argsStart, PTR_VASigCookie varArgSig, GCCONTEXT* ct
     }
 }
 
-INDEBUG(void* forceStack1;)
+#ifndef DACCESS_COMPILE
+// Verify the GC frame registration and the actual GCFrame are the same size.
+static_assert_no_msg(sizeof(GCFrameRegistration) == sizeof(GCFrame));
+
+FCIMPL1(void, GCReporting::Register, GCFrameRegistration* reg)
+{
+    FCALL_CONTRACT;
+
+    _ASSERTE(reg != NULL);
+    OBJECTREF* pObjRefs = reg->m_pObjRefs;
+    UINT32 numObjRefs = reg->m_numObjRefs;
+    BOOL maybeInterior = reg->m_MaybeInterior;
+
+    // Construct a GCFrame.
+    new (reg)GCFrame(pObjRefs, numObjRefs, maybeInterior);
```

Comment:
It would be nice to add `GCFrame::Push` method that will just add the struct to the linked list without reinitializing the already initialized fields.

Reply:
I've also added a `GCFrame::Pop` method; it made the registration logic clearer.